exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Query.kt

```diff
@@ -77,9 +87,43 @@ open class Query(override var set: FieldSet, where: Op<Boolean>?) : AbstractQuer
     }
 
     override fun withDistinct(value: Boolean): Query = apply {
+        if (value) {
+            require(distinctOn == null) { "DISTINCT cannot be used with the DISTINCT ON modifier. Only one of them should be applied." }
+        }
         distinct = value
     }
 
+    /**
+     * Specifies that the `SELECT` query should retrieve distinct results based on the given list of columns.
+     *
+     * This method can be used to set a `DISTINCT ON` clause for the query, which is supported by some SQL dialects
+     * (e.g., PostgreSQL, H2). The resulting query will retrieve rows that are distinct based on the specified columns.
+     *
+     * @param columns The columns to apply the `DISTINCT ON` clause.
+     * @return The current `Query` instance with the `DISTINCT ON` clause applied.
+     */
+    fun withDistinctOn(vararg columns: Column<*>): Query = apply {
+        require(!distinct) { "DISTINCT ON cannot be used with the DISTINCT modifier. Only one of them should be applied." }
+        distinctOn = (distinctOn ?: emptyList()) + columns
+    }
```

Comment:
The only uncovered case I see is if no values are (somehow) passed, like:
```kt
.withDistinctOn(columns = emptyArray<Column<*>>())
```

I see that `groupBy()` and `orderBy()` also allow no values to be passed at compile time but will ignore the clause in `prepareSQL()` because of a check for empty lists. `distinctOn` should probably have the same check, otherwise it causes a syntax exception.

I'd also think more about what should be done with the `require` in such a case when `columns` is empty:
1. Should behavior stay the same?
2. Should there be an early `return@apply`?

My first thought is 2, but there could be arguments for 1.

Reply:
I like the variant 2, because the call of `withDistinctOn` with empty list will not cause the change of query, and will not cause exception on `withDistinctCall`. Could allow to run code like:

```kotlin
val columns = arrayOf(...)
query
  .withDistinctOn(columns = columns)
  .withDistinct(columns.isEmpty())
```